src/os/bluestore/Allocator.h

```diff
@@ -16,8 +16,207 @@
 #include <ostream>
 #include "include/ceph_assert.h"
 #include "bluestore_types.h"
+#include "common/ceph_mutex.h"
+
+typedef interval_set<uint64_t> release_set_t;
+typedef release_set_t::value_type release_set_entry_t;
 
 class Allocator {
+protected:
+
+  struct ExtentCollectionTraits {
+    size_t num_buckets;
+    size_t base_bits; // min extent size
+    size_t base = 1ull << base_bits;
+    size_t factor;  // single bucket size range to be
+                    // determined as [len, len * factor * 2)
+                    // for log2(len) indexing and
+                    // [len, len + factor * base)
+                    // for linear indexing.
+
```

Comment:
The definition of buckets is unclear.
I am not sure what intended ranges are.
LinBucket [i] = [i * base, i * base + i * base * factor)  ?
ExpBucket[i] = [0 or previous-bucket-end, base * 2 ^ (factor * i) ) ?
I was unable to check _get_p2_size_bucket and _get_linear_size_bucket

Reply:
Hopefully clearer now..